lib/spack/spack/url.py

```diff
@@ -230,6 +230,9 @@ def parse_version_offset(path):
         # https://github.com/hpc/mpileaks/releases/download/v1.0/mpileaks-1.0.tar.gz
         (r'github.com/[^/]+/[^/]+/releases/download/v?([^/]+)/.*$', path),
 
+        # e.g., https://rubygems.org/downloads/rubysl-date-2.0.9.gem
+        (r'rubygems.org/.+/.*-((\d+\.)+\d+).gem$', path),
```

Comment:
This is now fixed in the latest develop branch so you can remove this change and rebase.
```
$ spack url parse https://rubygems.org/downloads/rubysl-date-2.0.9.gem
==> Parsing URL: https://rubygems.org/downloads/rubysl-date-2.0.9.gem

==> Matched version regex  0: r'^[a-zA-Z+._-]+[._-]v?(\\d[\\d._-]*)$'
==> Matched  name   regex  7: r'^([A-Za-z\\d+\\._-]+)$'

==> Detected:
    https://rubygems.org/downloads/rubysl-date-2.0.9.gem
                                   ----------- ~~~~~
    name:    rubysl-date
    version: 2.0.9

==> Substituting version 9.9.9b:
    https://rubygems.org/downloads/rubysl-date-9.9.9b.gem
                                   ----------- ~~~~~~
```